src/request.ts

```diff
@@ -89,6 +109,7 @@ class DatastoreRequest {
     | Array<(err: Error | null, resp: Entity | null) => void>
     | Entity;
   datastore!: Datastore;
+  protected state: TransactionState = TransactionState.NOT_TRANSACTION;
```

Comment:
I don't like using `protected` here because typescript doesn't technically hide `protected` properties completely, but it seems that there is no simple way around it because both `Transaction` and `DatastoreRequest` need access to the `state` variable. There is more information about some complex ways to avoid this in the design doc in Alternative 3, but they require major architecture changes.